Attempt to make it easier to detect when the request is done

davidfowl opened the pull request:
- Today the async local reference to the HttpContext flows when the execution context is captured. When the http request has ended, the HttpContext property will return the reference to an invalid HttpContext instead of returning null. This change stores both the request id and the HttpContext and makes sure both match before returning anything valid.
- This is still racy but should catch more cases of people doing bad things.
- There will still be issue if people store the context in a local and use that reference instead of accessing it through the property getter but we can live with that.
- Added tests

Mitigates https://github.com/aspnet/KestrelHttpServer/issues/2591

cc @benaadams

davidfowl committed:
Attempt to make it easier to detect when the request is done
- Today the async local reference to the HttpContext flows when the execution context is captured. When the http request has ended, the HttpContext property will return the reference to an invalid HttpContext instead of returning null. This change stores both the request id and the HttpContext and makes sure both match before returning anything valid.
- This is still racy but should catch more cases of people doing bad things.
- There will still be issue if people store the context in a local and use that reference instead of accessing it through the property getter but we can live with that.
- Added tests

davidfowl committed:
Added more tests

benaadams commented on src/Microsoft.AspNetCore.Http/HttpContextAccessor.cs:
```diff
@@ -7,17 +7,32 @@ namespace Microsoft.AspNetCore.Http
+                else
+                {
+                    // Setting the context to null means the request is over
+                    existing.context.TraceIdentifier = null;
```
Is set to `null` in `HttpProtocol.Reset()`?

davidfowl replied:
WebListener doesn't do this today so I'm not sure I want to remove it yet. It means all server implementations need to mutate the `HttpContext` on Dispose.

Tratcher replied:
That's better than having an arbitrary field mutated the accessor. Any way to give it its own field? A feature?

davidfowl replied:
I originally had that but it will make the performance worse (as it modifies the feature. We already have a feature that I'm piggy backing on. I would need to make a new feature and implement it in all the server feature collections then clear it here. It's exactly the same as this.

So the question is, what's the downside?

davidfowl committed:
PR feedback
- Set the TraceIdentifier to null in the default HttpContextFactory
- Added more tests

Tratcher commented on test/Microsoft.AspNetCore.Http.Tests/HttpContextAccessorTests.cs:
```diff
@@ -0,0 +1,197 @@
+        }
+
+        [Fact]
+        public async Task HttpContextAccessor_GettingHttpContextReturnsNullHttpContextIsDifferentTraceIdentifier()
```
IfDifferent?

Tratcher commented on src/Microsoft.AspNetCore.Http/HttpContextAccessor.cs:
```diff
@@ -7,17 +7,19 @@ namespace Microsoft.AspNetCore.Http
             {
-                return _httpContextCurrent.Value;
+                var value = _httpContextCurrent.Value;
+                // Only return the context if the stored request id matches the stored trace identifier
```
// TraceIdentifier is cleared by HttpContextFactory.Dispose.

davidfowl committed:
PR nits

halter73 commented on src/Microsoft.AspNetCore.Http/HttpContextAccessor.cs:
```diff
@@ -7,17 +7,20 @@ namespace Microsoft.AspNetCore.Http
             set
             {
-                _httpContextCurrent.Value = value;
+                _httpContextCurrent.Value = (value?.TraceIdentifier, value);
```
Is there ever a situation where a non-null HttpContext is set with a null TraceIdentifier? If not, this could be simplified quite a bit.

halter73 replied:
Nvm. I don't think it can be easily simplified.
